Giant puzzle #5, enconttrado no site da Giant Steps:
https://gscap.com.br/puzzle/

Giant Puzzle #5 – Blackjack
Zarathustra resolveu investigar uma oportunidade em um cassino quant que oferece variações do jogo de Blackjack. Em uma das variações disponíveis, você adiciona ao baralho 8 cartas I, com o valor de face contabilizado de maneira complexa.

A dinâmica da carta I é como uma exponenciação do complexo i e funciona da seguinte maneira:

Caso você tenha 1 carta I na mão, ela tem valor i, complexo, e não soma ao valor total, que é real, ou seja, 0 pontos;
Com duas cartas I o cenário muda de figura: I*I = i*i = -1, ou seja, você soma -1 pontos;
Com três cartas, voltamos ao cenário complexo: I3 = -i , ou seja, 0 pontos;
Finalmente, com quatro cartas: I4 = 1, ou seja, 1 ponto.
Considerando essa variação, qual o edge mínimo da casa com o jogador seguindo uma estratégia básica idêntica à do dealer, ou seja, escolhendo um valor fixo “soft” para parar de pedir novas cartas?

Extra
Sabemos que repetir a estratégia do dealer não é a melhor opção disponível.

Cite uma estratégia que aumente o valor esperado do jogador!

O ranking daqueles que solucionarem esse puzzle será baseado no valor esperado encontrado pela estratégia sugerida, sendo que aquele que submeteu a solução correta e o maior valor esperado estará no topo! Você pode reenviar sua solução caso encontre uma estratégia mais eficiente.

Obs.: Nesse jogo não é possível dobrar ou splitar e um baralho novo é utilizado a cada rodada. A estratégia do jogador não precisa parar em 17, pode ser qualquer valor. O edge da casa é a vantagem do dealer em relação ao jogador, ou seja, o lucro médio do dealer para cada real gasto pelo apostador – também podendo ser expresso em termos de porcentagem.

Regras do Blackjack
Blackjack é um dos jogos mais populares nos cassinos pelo mundo. O jogo é jogado entre um dealer contra um ou mais jogadores. Cada jogador recebe duas cartas enquanto o dealer apresenta uma carta virada para cima e outra para baixo. Cada carta vale o valor de face, sendo ‘J’=’Q’=’K’= 10 pontos e ‘A’=1 ou 11 pontos. A rodada é definida por cada jogador tomando uma das seguintes decisões:

Hit: pedir mais uma carta;
Stand: parar;
Dobrar: dobrar a aposta por mais uma carta;
Split: caso as duas primeiras cartas possuam o mesmo valor em pontos, é possível separá-las em dois jogos, onde cada uma dará início a uma nova mão. Uma aposta igual a inicial deve ser feita para a segunda mão.
O jogador ganha se a sua mão somar mais pontos que a do dealer, sem ir acima de 21 pontos (burst). No caso de empate, o jogador recebe de volta todo o valor da aposta, exceto quando o dealer tem um “Blackjack”(A e 10/J/Q/K) e o jogador não, nesse caso o jogador perde.

A estratégia do dealer é fixa: hit enquanto somar um total menor que 17 e stand para um total a partir de “soft” 17. Entenda “soft”como uma mão com um A em que ele é contabilizado como 11.

Em caso de vitória do jogador por “Blackjack” (A e 10/J/Q/K), ele recebe 2.5x a aposta, enquanto que em uma vitória simples ele recebe 2x a aposta inicial.


In [15]:
#!/usr/bin/python
# -*- coding: utf-8 -*-
#---------------------------------------------------------
from __future__ import print_function
import sys
import time
from yahoofinancials import YahooFinancials as yf

DEFAULT_ARGS = ('MMM', 'AAPL')
MODULE_ARGS = ('yf', 'yahoofinancial', 'yahoofinancials')
HELP_ARGS = ('-h', '--help')

mark = '-' * 64


import ipywidgets
import qgrid
import sys
from datetime import date

import numpy as np
import pandas as pd
from pandas import ExcelWriter
from pandas import ExcelFile
from bs4 import BeautifulSoup as bs
import requests
import random

In [16]:
valores = {"": 0, "A": 1, "2": 2, "3": 3, "4": 4, "5": 5, "6": 6, "7": 7, "8": 8, \
           "9": 9, "D": 10, "J": 10, "Q": 10, "K": 10, "I": 1j}
quant = [4,4,4,4,4,4,4,4,4,4,4,4,4,8]
# baralho = "AAAA22223333444455556666777788889999DDDDJJJJQQQQKKKKIIIIIIII"

btam = len(baralho)

len(quant)
len(valores)
len(baralho)

60

In [17]:
1j*1j*1j*1j

(1-0j)

In [18]:
random.seed(a=None, version=2)

it_max = 100

for i in range(it_max):
    val = random.randrange(0,btam)
#     print(val)

In [41]:
class jogador:
    
    # inicialização da classe
    #--------------------------------------
    def __init__(self, name, soft, baralho="", my_cards="", adv_cards=""):
    
        random.seed(a=None, version=2)
        
        self.name = name
        self.adv_cards = adv_cards
#         self.adv_p = valores[self.adv_card]

        self.soft = soft
        self.baralho = baralho
        
        self.seq_re = ""
        self.seq_img = ""
        self.seq = my_cards
        
        
        for my_card in my_cards:
            if my_card == "I":
                self.seq_img+= my_card
            else:
                self.seq_re += my_card
        
        self.points = self.conta_pontos(self.seq_re, self.seq_img)
    #--------------------------------------
    # setters
    #--------------------------------------
    def setBaralho(self, baralho):
        self.baralho = baralho
    
    # getters
    #--------------------------------------
    def getBaralho(self):
        return self.baralho
    #--------------------------------------
    def getSeq(self):
        return self.seq
    #--------------------------------------
    def getPoints(self):
        return self.points
    #--------------------------------------
    def getSoft(self):
        return self.soft
    #--------------------------------------
    def getAdvCard(self):
        return self.adv_cards
    #--------------------------------------
    def getAdvPoints(self):
        return self.adv_p
    
    # funções de jogo
    #--------------------------------------
    def isBlackjack(self):
        
        seq1 = "ADAJAQAK"
        seq2 = "DAJAQAKA"
        if (len(self.seq) == 2) and (self.seq in seq1 or self.seq in seq2):
            return True
        else:
            return False
            
    #--------------------------------------
    def hit(self):
        i = random.randrange(0,len(self.baralho))
        card = self.baralho[i]
        self.baralho = self.baralho[0:i] + self.baralho[i+1:len(self.baralho)]
        return card
    
    #--------------------------------------
    def conta_pontos(self, seq_re, seq_img):
        
        p_re = 0
        p_img = 0
        
        if seq_img != "":   
            p_img = 1j**len(seq_img)
        
        for card in seq_re:
            
            if card == "A":
                if p_re+p_img.real+11 >= self.soft and p_re+p_img.real+11 <= 21:
                    p_re += 11
                else:
                    p_re += 1
            else:    
                p_re += valores[card]
        
        
        p = p_re + p_img
        
        return p.real
    #--------------------------------------
    
    def play(self):
        
        self.points = self.conta_pontos(self.seq_re, self.seq_img)
        
        while self.points < self.soft:
            card = self.hit()
            
            self.seq += card
            if card == "I":
                self.seq_img += card
            else:
                self.seq_re += card
                
            self.points = self.conta_pontos(self.seq_re, self.seq_img)
            
    #--------------------------------------
    def print(self):
        print("Pontos " + self.name + " = ", self.getPoints())
        print("Sequência "+ self.name + " = ", self.getSeq())
    
    #--------------------------------------
        

In [178]:
p = 1e6
n_max = 1000000
aposta = 1

soft_dealer = 17
soft_jogador = 15

p_dealer_ini = p
p_dealer = p

p_jogador_ini = p
p_jogador = p

# função que realiza a rodada:
#------------------------------------
def rodada(dealer, j):
    
    global p_dealer
    global p_jogador
    global aposta

    
    p_jogador -= aposta
    
    j.play()
    baralho = j.getBaralho()
    dealer.setBaralho(baralho)
    dealer.play()
    
    pd = dealer.getPoints()
    pj = j.getPoints()
    
    if pd > 21:
        pd = -1
    if pj > 21:
        pj = -1
        
#     if pj == -1 and pd == -1:
#         p_jogador += aposta
    if pj == -1:
        p_dealer += aposta
        
    elif pj > pd and j.isBlackjack() == True:
        p_jogador += 2.5* aposta
        p_dealer -= 1.5*aposta
        
#     elif pj > pd and pj <= 21:
    elif pj > pd:
        p_jogador += 2*aposta
        p_dealer -= aposta
        
    elif pj == pd:
        if dealer.isBlackjack() == True and j.isBlackjack() == False:
            p_dealer += aposta
        else:
            p_jogador += aposta
    
#     elif pd >= pj:
#         p_dealer += aposta
    
    else:
        p_dealer += aposta

#------------------------------------  
def hit(baralho):
        random.seed(a=None, version=2)
        i = random.randrange(0,len(baralho))
        card = baralho[i]
        baralho = baralho[0:i] + baralho[i+1:len(baralho)]
#         print("i = ", i)
#         print("baralho = ", baralho)
        return card, baralho
#------------------------------------    


for i in range(n_max):
    
    baralho = "AAAA22223333444455556666777788889999DDDDJJJJQQQQKKKKIIIIIIII"
#     baralho = "AAAA22223333444455556666777788889999DDDDJJJJQQQQKKKK"
    
    dealer_card1, baralho = hit(baralho)
    dealer_card2, baralho = hit(baralho)
    
    dealer_cards = dealer_card1 + dealer_card2
    
    j1_card1, baralho = hit(baralho)
    j1_card2, baralho = hit(baralho)
    
    j1_cards = j1_card1 + j1_card2    
    
    dealer = jogador("dealer", soft_dealer, baralho, dealer_cards, j1_cards)
    j1 = jogador("jogador", soft_jogador, baralho, j1_cards, dealer_cards)
    rodada(dealer, j1)

    
# dealer.print()
# j1.print()

print("Dinheiro do dealer = ", p_dealer)
print("Dinheiro do jogador = ", p_jogador)

edge = (p_dealer - p_dealer_ini) / (n_max*aposta)

print("Edge do dealer = ", edge*100, "%")
# print("baralho = ", j1.getBaralho())
# print("len baralho = ", len(j1.getBaralho()))


Dinheiro do dealer =  1065350.5
Dinheiro do jogador =  934649.5
Edge do dealer =  6.535050000000001 %


resultado com soft_jogador 10:
n_max = 100 mil iterações
Dinheiro do dealer =  1016448.0
Dinheiro do jogador =  983552.0
Edge do dealer =  16.448 %

resultado com soft_jogador 11:
n_max = 100 mil iterações
Dinheiro do dealer =  1009841.5
Dinheiro do jogador =  990158.5
Edge do dealer =  9.8415 %

resultado com soft_jogador 12:
n_max = 100 mil iterações
Dinheiro do dealer =  1009115.0
Dinheiro do jogador =  990885.0
Edge do dealer =  9.115 %

resultado com soft_jogador 13:
n_max = 100 mil iterações
Dinheiro do dealer =  1006816.5
Dinheiro do jogador =  993183.5
Edge do dealer =  6.8165000000000004 %

resultado com soft_jogador 14:
n_max = 100 mil iterações
Dinheiro do dealer =  1006500.0
Dinheiro do jogador =  993500.0
Edge do dealer =  6.5 %

resultado com soft_jogador 14:
n_max = 1 milhão de iterações
Dinheiro do dealer =  1065690.0
Dinheiro do jogador =  934310.0
Edge do dealer =  6.569 %

resultado com soft_jogador = 15:
n_max = 100 mil iterações
Dinheiro do dealer =  1006465.5
Dinheiro do jogador =  993534.5
Edge do dealer =  6.4655000000000005 %

resultado com soft_jogador = 15:
n_max = 1 milhão de iterações
Dinheiro do dealer =  1065350.5
Dinheiro do jogador =  934649.5
Edge do dealer =  6.535050000000001 %

resultado com soft_jogador = 16:
n_max = 100 mil iterações
Dinheiro do dealer =  1007622.5
Dinheiro do jogador =  992377.5
Edge do dealer =  7.6225000000000005 %

resultado com soft_jogador = 17:
n_max = 100 mil iterações
Dinheiro do dealer =  1007898.5
Dinheiro do jogador =  992101.5
Edge do dealer =  7.8985 %

resultado com soft_jogador = 18:
n_max = 100 mil iterações
Dinheiro do dealer =  1013202.5
Dinheiro do jogador =  986797.5
Edge do dealer =  13.2025 %


In [166]:
# função que realiza a rodada:
#------------------------------------
def rodada(dealer, j):
    
    global p_dealer
    global p_jogador
    global aposta

    global lost_hand_count
    global busts_count
    global total_count
    global empate_count
    
    total_count += 1
    
    p_jogador -= aposta
    
    j.play()
    baralho = j.getBaralho()
    dealer.setBaralho(baralho)
    dealer.play()
    
    pd = dealer.getPoints()
    pj = j.getPoints()
    
    if pd > 21:
        pd = -1
    if pj > 21:
        pj = -1
        
#     if pj == -1 and pd == -1:
#         p_jogador += aposta
    if pj == -1:
        p_dealer += aposta
        busts_count += 1
        
    elif pj > pd and j.isBlackjack() == True:
        p_jogador += 2.5* aposta
        p_dealer -= 1.5*aposta
        
#     elif pj > pd and pj <= 21:
    elif pj > pd:
        p_jogador += 2*aposta
        p_dealer -= aposta
        
    elif pj == pd:
        if dealer.isBlackjack() == True and j.isBlackjack() == False:
            p_dealer += aposta
            lost_hand_count +=1
        else:
            p_jogador += aposta
            empate_count += 1
    
#     elif pd >= pj:
#         p_dealer += aposta
    
    else:
        p_dealer += aposta
        lost_hand_count +=1

#------------------------------------  
def hit(baralho):
    random.seed(a=None, version=2)
    i = random.randrange(0,len(baralho))
    card = baralho[i]
    baralho = baralho[0:i] + baralho[i+1:len(baralho)]
#         print("i = ", i)
#         print("baralho = ", baralho)
    return card, baralho
#------------------------------------  
def select(baralho, card):
    baralho = baralho[0:baralho == card] + baralho[(baralho == card)+1: len(baralho)]
    return card, baralho

#------------------------------------    
# função que realiza o jogo completo
def game(p, n_max, soft_dealer, soft_jogador, carta_dealer):

    global p_dealer
    global p_jogador
    global aposta
    
    global lost_hand_count
    global busts_count
    global total_count
    global empate_count
    
    lost_hand_count = 0
    busts_count = 0
    total_count = 0
    empate_count = 0
    
    p_dealer_ini = p
    p_dealer = p

    p_jogador_ini = p
    p_jogador = p

    for i in range(n_max):


        baralho = "AAAA22223333444455556666777788889999DDDDJJJJQQQQKKKKIIIIIIII"
    #     baralho = "AAAA22223333444455556666777788889999DDDDJJJJQQQQKKKK"

        dealer_card1, baralho = select(baralho, carta_dealer)
        dealer_card2, baralho = hit(baralho)

        dealer_cards = dealer_card1 + dealer_card2

        j1_card1, baralho = hit(baralho)
        j1_card2, baralho = hit(baralho)

        j1_cards = j1_card1 + j1_card2    

        dealer = jogador("dealer", soft_dealer, baralho, dealer_cards, j1_cards)
        j1 = jogador("jogador", soft_jogador, baralho, j1_cards, dealer_cards)
        rodada(dealer, j1)


    # dealer.print()
    # j1.print()
    edge = 100 * (p_dealer - p_dealer_ini) / (n_max*aposta)
    
#     print("Resultado com soft joagador = ", soft_jogador, "e carta do dealer = ", carta_dealer)
#     print("Quantidade de iterações = ", n_max)
#     print("Dinheiro do dealer = ", p_dealer)
#     print("Dinheiro do jogador = ", p_jogador)
#     print("Edge do dealer = ", edge, "%")
    # print("baralho = ", j1.getBaralho())
    # print("len baralho = ", len(j1.getBaralho()))
    
    return edge

#------------------------------------

p = 1e6
n_max = 10000
aposta = 1
min_soft_jogador = 10
soft_dealer = 17
list_soft_jogador = list(range(min_soft_jogador,22))

cartas = "A23456789DJQKI"

results = np.array(np.zeros([len(cartas), len(list_soft_jogador)]))
lost_hand = np.array(np.zeros([len(cartas), len(list_soft_jogador)]))
busts = np.array(np.zeros([len(cartas), len(list_soft_jogador)]))
empate = np.array(np.zeros([len(cartas), len(list_soft_jogador)]))
total = np.array(np.zeros([len(cartas), len(list_soft_jogador)]))

# posicao = {"A": 0, "2": 1, "3": 2, "4": 3, "5": 4, "6": 5, "7": 6, "8": 7, "9": 8, "D": 9, "J": 9, "Q": 9, "K": 9, "I": 10}

for carta_dealer in cartas:
    for soft_jogador in list_soft_jogador:
        edge = game(p, n_max, soft_dealer, soft_jogador, carta_dealer)
        results[cartas.index(carta_dealer)][soft_jogador - min_soft_jogador]= edge
        lost_hand[cartas.index(carta_dealer)][soft_jogador - min_soft_jogador]= lost_hand_count
        busts[cartas.index(carta_dealer)][soft_jogador - min_soft_jogador]= busts_count
        empate[cartas.index(carta_dealer)][soft_jogador - min_soft_jogador]= empate_count
        total[cartas.index(carta_dealer)][soft_jogador - min_soft_jogador]= total_count
        

print("Quantidade de iterações = ", n_max)
np.set_printoptions(precision=3)

df_results = pd.DataFrame(results,index=list(cartas), columns=list_soft_jogador)
print("Edge do dealer")
display(df_results)

df_total = pd.DataFrame(total,index=list(cartas), columns=list_soft_jogador)
print("Total")
display(df_total)

df_busts = pd.DataFrame(100*busts/total,index=list(cartas), columns=list_soft_jogador)
print("Busts")
display(df_busts)

df_lost_hand = pd.DataFrame(100*lost_hand/total,index=list(cartas), columns=list_soft_jogador)
print("Perdas")
display(df_lost_hand)

df_empate = pd.DataFrame(100*empate/total,index=list(cartas), columns=list_soft_jogador)
print("Empate")
display(df_empate)

df_lost_hand_plus_busts = pd.DataFrame(100*(lost_hand/total + busts / total),index=list(cartas), columns=list_soft_jogador)
print("Perdas + Busts")
display(df_lost_hand_plus_busts)


Quantidade de iterações =  10000
Edge do dealer


10      11      12      13      14      15      16      17      18  \
A  18.675  13.905  10.915  11.040  10.325  10.140  12.175  12.865  16.485   
2   7.140   0.730  -1.545  -0.645  -0.610  -0.840   1.855   3.820   9.135   
3   3.560  -1.585  -3.985  -5.095  -3.465  -2.670  -0.485   4.815   8.835   
4   0.380  -3.130  -6.350  -8.275  -6.080  -4.470  -1.835   2.415   9.900   
5  -1.470  -7.265  -8.995 -11.285  -8.780  -8.200  -4.430   1.195   7.470   
6  -6.280  -9.550 -12.390 -11.835 -12.525  -8.675  -6.560  -1.440   5.955   
7  13.305   6.920   3.520  -0.300  -1.195  -3.875  -4.530  -5.760   1.120   
8  23.205  13.730  10.040   8.155   4.360   3.380   2.625   0.250   3.895   
9  29.000  22.490  18.665  13.455  12.555  13.665  10.800  11.255  11.680   
D  35.275  28.335  26.160  24.880  24.180  23.370  21.705  23.125  23.685   
J  37.365  28.830  27.925  24.610  23.415  23.210  22.305  22.205  24.680   
Q  34.385  29.530  26.545  26.610  25.260  23.625  22.825  21.905  25.040   
K  35.000  30.690  28.245  25.530  22.515  22.255  23.245  23.160  23.840   
I  17.605  11.370   9.660   7.630   6.610   4.735   6.745   6.750  11.435   

       19      20      21  
A  27.070  41.480  71.935  
2  21.190  39.735  70.080  
3  22.300  38.995  71.155  
4  20.810  39.930  71.080  
5  18.835  37.820  69.565  
6  19.180  38.445  72.165  
7  15.350  36.465  69.455  
8  16.485  37.865  69.520  
9  19.705  36.505  70.990  
D  30.720  43.510  71.125  
J  31.755  43.830  71.060  
Q  30.860  42.845  72.310  
K  31.895  42.310  70.675  
I  22.715  38.620  72.235

Total


10       11       12       13       14       15       16       17  \
A  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
2  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
3  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
4  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
5  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
6  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
7  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
8  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
9  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
D  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
J  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
Q  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
K  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
I  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   

        18       19       20       21  
A  10000.0  10000.0  10000.0  10000.0  
2  10000.0  10000.0  10000.0  10000.0  
3  10000.0  10000.0  10000.0  10000.0  
4  10000.0  10000.0  10000.0  10000.0  
5  10000.0  10000.0  10000.0  10000.0  
6  10000.0  10000.0  10000.0  10000.0  
7  10000.0  10000.0  10000.0  10000.0  
8  10000.0  10000.0  10000.0  10000.0  
9  10000.0  10000.0  10000.0  10000.0  
D  10000.0  10000.0  10000.0  10000.0  
J  10000.0  10000.0  10000.0  10000.0  
Q  10000.0  10000.0  10000.0  10000.0  
K  10000.0  10000.0  10000.0  10000.0  
I  10000.0  10000.0  10000.0  10000.0

Busts


10   11    12    13    14     15     16     17     18     19     20     21
A  0.0  0.0  0.00  3.72  8.28  14.50  22.30  30.91  40.70  52.05  64.71  84.93
2  0.0  0.0  0.00  4.01  8.71  14.39  21.06  30.03  39.70  51.85  65.81  84.32
3  0.0  0.0  0.00  3.95  8.66  14.61  21.77  30.79  40.06  52.71  65.34  85.00
4  0.0  0.0  0.00  3.64  8.35  14.43  22.30  31.13  40.79  51.84  66.15  85.13
5  0.0  0.0  0.00  3.56  8.59  14.18  21.99  30.15  40.10  51.59  65.29  84.28
6  0.0  0.0  0.00  3.72  8.04  14.61  22.04  30.90  40.26  52.19  65.42  85.69
7  0.0  0.0  0.00  3.55  8.60  14.69  21.98  30.29  40.51  51.97  65.52  84.50
8  0.0  0.0  0.00  3.77  8.46  14.77  21.51  29.95  40.73  52.91  66.53  84.53
9  0.0  0.0  0.00  3.82  8.02  14.55  21.46  29.96  40.22  52.20  65.77  85.38
D  0.0  0.0  0.00  3.37  8.46  14.23  21.59  30.44  39.85  51.89  65.59  84.83
J  0.0  0.0  0.00  3.87  8.21  14.38  21.76  29.86  40.17  51.84  65.94  84.80
Q  0.0  0.0  0.00  3.84  8.99  14.66  21.81  30.69  40.57  52.05  65.45  85.44
K  0.0  0.0  0.01  3.54  7.76  14.12  21.61  30.37  39.49  52.32  64.89  84.68
I  0.0  0.0  0.00  3.56  8.73  14.71  21.48  30.05  40.37  52.13  65.47  85.56

Perdas


10     11     12     13     14     15     16     17     18    19    20  \
A  57.65  54.79  52.67  48.73  43.49  36.89  29.43  20.85  13.84  8.17  3.58   
2  52.05  48.45  46.76  42.67  38.13  31.72  26.15  17.86  11.28  6.14  2.51   
3  50.35  47.25  45.61  40.79  36.67  30.86  24.34  17.83  10.96  5.91  2.44   
4  49.10  46.53  44.36  39.60  35.51  30.02  23.13  16.01  11.15  6.00  2.25   
5  47.96  44.50  43.18  38.26  34.25  28.63  22.46  16.48  10.43  5.45  2.14   
6  45.39  43.31  41.67  37.90  32.73  28.08  21.36  14.93   9.70  5.16  2.21   
7  54.64  51.00  48.73  42.67  36.87  29.18  21.32  11.81   7.54  4.17  1.53   
8  59.69  54.44  51.97  46.44  40.26  32.64  25.19  15.45   6.78  3.59  1.37   
9  62.71  58.75  56.37  49.58  44.40  37.94  29.35  20.55  10.96  3.95  1.29   
D  65.73  61.26  59.59  55.19  49.49  43.02  34.43  25.99  17.31  9.12  2.59   
J  66.78  61.74  60.64  54.96  49.36  43.18  34.67  26.00  17.21  9.58  2.64   
Q  65.15  61.93  59.82  55.85  49.85  42.79  34.96  25.23  17.47  9.29  2.60   
K  65.64  62.68  60.88  55.90  49.76  42.53  35.19  26.11  17.60  9.63  2.70   
I  57.23  53.53  52.05  47.15  41.10  33.78  27.66  18.87  11.64  6.07  1.80   

     21  
A  0.00  
2  0.00  
3  0.00  
4  0.00  
5  0.00  
6  0.00  
7  0.00  
8  0.00  
9  0.00  
D  0.67  
J  0.70  
Q  0.58  
K  0.57  
I  0.00

Empate


10    11    12    13    14    15     16     17     18    19    20    21
A  4.47  5.51  6.15  6.69  7.40  7.94   9.27   9.95   8.04  7.16  5.46  2.64
2  4.19  5.10  5.48  6.60  6.43  7.55   8.15   8.64   7.76  5.85  3.60  2.02
3  4.14  5.20  5.42  6.04  6.59  6.97   7.94   8.21   7.33  5.62  4.04  1.76
4  3.41  5.07  5.56  6.02  6.85  7.23   7.95   8.65   6.55  5.75  3.74  1.46
5  3.86  4.96  5.27  5.63  6.17  6.80   7.42   8.52   7.01  5.51  3.57  1.70
6  4.26  5.00  4.92  5.57  6.60  6.66   7.36   7.56   6.75  5.14  3.85  1.29
7  5.32  6.14  6.73  7.87  8.50  9.04   9.59  10.61   5.64  3.73  2.98  1.11
8  5.17  6.16  6.71  8.40  7.62  9.16   9.83  10.14   9.43  4.08  2.61  1.00
9  4.91  6.19  6.58  7.38  8.51  9.35   9.86  10.97   9.95  8.11  3.05  0.88
D  5.11  7.10  7.57  8.41  8.98  9.49  10.33  10.88  10.06  9.37  7.83  0.82
J  5.09  6.57  7.18  7.58  8.91  8.83  10.13  11.18  10.48  9.52  7.31  0.69
Q  5.36  6.88  7.51  7.72  8.28  9.38   9.88  10.76   9.58  8.75  7.40  0.83
K  5.06  6.55  7.18  7.28  8.10  9.70  10.29  10.84  10.29  8.76  7.82  0.75
I  4.30  5.60  6.20  6.88  7.63  8.33   9.03   9.65   8.11  6.92  4.69  1.65

Perdas + Busts


10     11     12     13     14     15     16     17     18     19  \
A  57.65  54.79  52.67  52.45  51.77  51.39  51.73  51.76  54.54  60.22   
2  52.05  48.45  46.76  46.68  46.84  46.11  47.21  47.89  50.98  57.99   
3  50.35  47.25  45.61  44.74  45.33  45.47  46.11  48.62  51.02  58.62   
4  49.10  46.53  44.36  43.24  43.86  44.45  45.43  47.14  51.94  57.84   
5  47.96  44.50  43.18  41.82  42.84  42.81  44.45  46.63  50.53  57.04   
6  45.39  43.31  41.67  41.62  40.77  42.69  43.40  45.83  49.96  57.35   
7  54.64  51.00  48.73  46.22  45.47  43.87  43.30  42.10  48.05  56.14   
8  59.69  54.44  51.97  50.21  48.72  47.41  46.70  45.40  47.51  56.50   
9  62.71  58.75  56.37  53.40  52.42  52.49  50.81  50.51  51.18  56.15   
D  65.73  61.26  59.59  58.56  57.95  57.25  56.02  56.43  57.16  61.01   
J  66.78  61.74  60.64  58.83  57.57  57.56  56.43  55.86  57.38  61.42   
Q  65.15  61.93  59.82  59.69  58.84  57.45  56.77  55.92  58.04  61.34   
K  65.64  62.68  60.89  59.44  57.52  56.65  56.80  56.48  57.09  61.95   
I  57.23  53.53  52.05  50.71  49.83  48.49  49.14  48.92  52.01  58.20   

      20     21  
A  68.29  84.93  
2  68.32  84.32  
3  67.78  85.00  
4  68.40  85.13  
5  67.43  84.28  
6  67.63  85.69  
7  67.05  84.50  
8  67.90  84.53  
9  67.06  85.38  
D  68.18  85.50  
J  68.58  85.50  
Q  68.05  86.02  
K  67.59  85.25  
I  67.27  85.56

In [167]:
print("Quantidade de iterações = ", n_max)

print("Edge do dealer")
display(pd.DataFrame(df_results.min(axis=1)).transpose())
display(pd.DataFrame(df_results.idxmin(axis=1)).transpose())

print("Perdas + Busts")
display(pd.DataFrame(df_lost_hand_plus_busts.min(axis=1)).transpose())
display(pd.DataFrame(df_lost_hand_plus_busts.idxmin(axis=1)).transpose())

print("Total")
display(pd.DataFrame(df_total.min(axis=1)).transpose())
display(pd.DataFrame(df_total.idxmin(axis=1)).transpose())

print("Busts")
display(pd.DataFrame(df_busts.min(axis=1)).transpose())
display(pd.DataFrame(df_busts.idxmin(axis=1)).transpose())

print("Perdas")
display(pd.DataFrame(df_lost_hand.min(axis=1)).transpose())
display(pd.DataFrame(df_lost_hand.idxmin(axis=1)).transpose())

print("Empate")
display(pd.DataFrame(df_empate.min(axis=1)).transpose())
display(pd.DataFrame(df_empate.idxmin(axis=1)).transpose())


Quantidade de iterações =  10000
Edge do dealer


A      2      3      4       5       6     7     8     9       D  \
0  10.14 -1.545 -5.095 -8.275 -11.285 -12.525 -5.76  0.25  10.8  21.705   

        J       Q       K      I  
0  22.205  21.905  22.255  4.735

A   2   3   4   5   6   7   8   9   D   J   Q   K   I
0  15  12  13  13  13  14  17  17  16  16  17  17  15  15

Perdas + Busts


A      2      3      4      5      6     7     8      9      D      J  \
0  51.39  46.11  44.74  43.24  41.82  40.77  42.1  45.4  50.51  56.02  55.86   

       Q      K      I  
0  55.92  56.48  48.49

A   2   3   4   5   6   7   8   9   D   J   Q   K   I
0  15  15  13  13  13  14  17  17  17  16  17  17  17  15

Total


A        2        3        4        5        6        7        8  \
0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   

         9        D        J        Q        K        I  
0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0

A   2   3   4   5   6   7   8   9   D   J   Q   K   I
0  10  10  10  10  10  10  10  10  10  10  10  10  10  10

Busts


A    2    3    4    5    6    7    8    9    D    J    Q    K    I
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0

A   2   3   4   5   6   7   8   9   D   J   Q   K   I
0  10  10  10  10  10  10  10  10  10  10  10  10  10  10

Perdas


A    2    3    4    5    6    7    8    9     D    J     Q     K    I
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.67  0.7  0.58  0.57  0.0

A   2   3   4   5   6   7   8   9   D   J   Q   K   I
0  21  21  21  21  21  21  21  21  21  21  21  21  21  21

Empate


A     2     3     4    5     6     7    8     9     D     J     Q     K  \
0  2.64  2.02  1.76  1.46  1.7  1.29  1.11  1.0  0.88  0.82  0.69  0.83  0.75   

      I  
0  1.65

A   2   3   4   5   6   7   8   9   D   J   Q   K   I
0  21  21  21  21  21  21  21  21  21  21  21  21  21  21

In [168]:
dic_lost_hand_plus_busts = pd.DataFrame(df_lost_hand_plus_busts.idxmin(axis=1)).to_dict()[0]
dic_edge_results = pd.DataFrame(df_results.idxmin(axis=1)).to_dict()[0]
print(dic_lost_hand_plus_busts)
print(dic_edge_results)

{'A': 15, '2': 15, '3': 13, '4': 13, '5': 13, '6': 14, '7': 17, '8': 17, '9': 17, 'D': 16, 'J': 17, 'Q': 17, 'K': 17, 'I': 15}
{'A': 15, '2': 12, '3': 13, '4': 13, '5': 13, '6': 14, '7': 17, '8': 17, '9': 16, 'D': 16, 'J': 17, 'Q': 17, 'K': 15, 'I': 15}


## Tabela usada para determinar o soft de parada flutuante do jogador.
Essa tabela foi obtida por meio das iterações realizadas na célula anterior.

|   |     10 |     11 |      12 |      13 |      14 |     15 |     16 |     17 |     18 |     19 |     20 |     21 |
|--:|-------:|-------:|--------:|--------:|--------:|-------:|-------:|-------:|-------:|-------:|-------:|-------:|
| A | 18.385 | 12.455 |  12.310 |   9.105 |   9.580 | 11.180 | 10.755 | 13.005 | 15.185 | 25.710 | 42.350 | 72.000 |
| 2 |  7.675 |  1.740 |  -2.010 |  -3.120 |  -1.730 |  0.490 |  3.175 |  3.825 | 10.705 | 21.235 | 39.935 | 70.195 |
| 3 |  2.045 | -2.930 |  -4.115 |  -6.090 |  -5.200 | -2.860 | -2.015 |  3.730 |  9.155 | 21.225 | 39.670 | 70.315 |
| 4 | -1.215 | -6.150 |  -6.165 |  -6.455 |  -8.525 | -5.065 | -1.805 |  0.415 | 10.160 | 20.185 | 39.125 | 70.140 |
| 5 | -3.915 | -8.295 | -10.950 |  -9.905 | -10.205 | -7.870 | -3.910 |  0.265 |  6.270 | 18.545 | 40.835 | 71.095 |
| 6 | -5.665 | -9.475 | -14.040 | -13.775 | -12.225 | -9.620 | -4.940 | -1.760 |  7.115 | 20.125 | 37.945 | 71.080 |
| 7 | 13.860 |  6.290 |   3.290 |   0.775 |  -2.770 | -5.670 | -4.100 | -5.210 | -1.410 | 14.740 | 36.840 | 70.085 |
| 8 | 21.000 | 12.500 |   9.530 |   7.360 |   5.230 |  3.815 |  2.785 |  1.980 |  4.485 | 13.750 | 36.255 | 70.615 |
| 9 | 28.055 | 20.525 |  16.470 |  16.325 |  15.050 | 13.570 | 12.300 | 11.415 | 12.145 | 19.970 | 36.075 | 70.335 |
| D | 35.030 | 29.215 |  26.210 |  25.185 |  23.870 | 22.680 | 23.220 | 22.575 | 23.555 | 33.275 | 43.090 | 71.515 |
| I | 16.980 |  9.375 |   9.770 |   7.675 |   6.405 |  6.965 |  6.890 |  7.630 | 12.495 | 22.670 | 39.385 | 69.225 |

In [174]:
# Dicionário a ser utilizado para os testes com a estratégia de soft de parada flutuante do jogador
# a simulação para esse caso é realizada na célula seguinte.

# dic_soft_jogador = {"A": 13, "2": 13, "3": 13, "4": 14, "5": 12, "6": 12, "7": 15, \
#                     "8": 17, "9": 17, "D": 17, "J": 17, "Q": 17, "K": 17, "I": 14}

# dic_soft_jogador = {"A": 13, "2": 11, "3": 10, "4": 10, "5": 10, "6": 10, "7": 11, \
#                     "8": 14, "9": 17, "D": 17, "J": 17, "Q": 17, "K": 17, "I": 14}
dic_lost_hand_plus_busts = pd.DataFrame(df_lost_hand_plus_busts.idxmin(axis=1)).to_dict()[0]
dic_edge_results = pd.DataFrame(df_results.idxmin(axis=1)).to_dict()[0]

dic_soft_jogador = dic_lost_hand_plus_busts
# dic_soft_jogador = dic_edge_results
print(dic_soft_jogador)

{'A': 15, '2': 15, '3': 13, '4': 13, '5': 13, '6': 14, '7': 17, '8': 17, '9': 17, 'D': 16, 'J': 17, 'Q': 17, 'K': 17, 'I': 15}


In [173]:
p = 1e6
n_max = 100000
aposta = 1

soft_dealer = 17
soft_jogador = 15

p_dealer_ini = p
p_dealer = p

p_jogador_ini = p
p_jogador = p

# função que realiza a rodada:
#------------------------------------
def rodada(dealer, j):
    
    global p_dealer
    global p_jogador
    global aposta

    
    p_jogador -= aposta
    
    j.play()
    baralho = j.getBaralho()
    dealer.setBaralho(baralho)
    dealer.play()
    
    pd = dealer.getPoints()
    pj = j.getPoints()
    
    if pd > 21:
        pd = -1
    if pj > 21:
        pj = -1
        
#     if pj == -1 and pd == -1:
#         p_jogador += aposta
    if pj == -1:
        p_dealer += aposta
        
    elif pj > pd and j.isBlackjack() == True:
        p_jogador += 2.5* aposta
        p_dealer -= 1.5*aposta
        
#     elif pj > pd and pj <= 21:
    elif pj > pd:
        p_jogador += 2*aposta
        p_dealer -= aposta
        
    elif pj == pd:
        if dealer.isBlackjack() == True and j.isBlackjack() == False:
            p_dealer += aposta
        else:
            p_jogador += aposta
    
#     elif pd >= pj:
#         p_dealer += aposta
    
    else:
        p_dealer += aposta

#------------------------------------  
def hit(baralho):
        random.seed(a=None, version=2)
        i = random.randrange(0,len(baralho))
        card = baralho[i]
        baralho = baralho[0:i] + baralho[i+1:len(baralho)]
#         print("i = ", i)
#         print("baralho = ", baralho)
        return card, baralho
#------------------------------------    


for i in range(n_max):
    
    baralho = "AAAA22223333444455556666777788889999DDDDJJJJQQQQKKKKIIIIIIII"
#     baralho = "AAAA22223333444455556666777788889999DDDDJJJJQQQQKKKK"
    
    dealer_card1, baralho = hit(baralho)
    dealer_card2, baralho = hit(baralho)
    
    dealer_cards = dealer_card1 + dealer_card2
    
    j1_card1, baralho = hit(baralho)
    j1_card2, baralho = hit(baralho)
    
    j1_cards = j1_card1 + j1_card2    
    
    dealer = jogador("dealer", soft_dealer, baralho, dealer_cards, j1_cards)
    soft_jogador = dic_soft_jogador[dealer_card1]
    j1 = jogador("jogador", soft_jogador, baralho, j1_cards, dealer_cards)
    rodada(dealer, j1)

    
edge = 100*(p_dealer - p_dealer_ini) / (n_max*aposta)

print("Quantidade de iterações = ", n_max)
print("Dinheiro do dealer = ", p_dealer)
print("Dinheiro do jogador = ", p_jogador)
print("Edge do dealer = ", edge, "%")


Quantidade de iterações =  100000
Dinheiro do dealer =  1005265.5
Dinheiro do jogador =  994734.5
Edge do dealer =  5.2655 %


In [ ]:
baralho = "AAAA22223333444455556666777788889999DDDDJJJJQQQQKKKKIIIIIIII"

valores = {"": 0, "A": 1, "2": 2, "3": 3, "4": 4, "5": 5, "6": 6, "7": 7, "8": 8, \
           "9": 9, "D": 10, "J": 10, "Q": 10, "K": 10, "I": 1j}

def completa_cartas()

p = 1e6
n_max = 100000
aposta = 1

soft_dealer = 17
soft_jogador = 15

p_dealer_ini = p
p_dealer = p

p_jogador_ini = p
p_jogador = p


j1 = jogador("jogador", soft_jogador, baralho, j1_cards, dealer_cards)